fix(elements): detect matchesSelector prototype without IIFE

CaerusKaru opened the pull request:
Although in SSR we patch the global prototypes with DOM globals
like Element and Node, this patch does not occur before the
matches function is called in Angular Elements. This is similar
to the behavior in @angular/upgrade.

Fixes #24551

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: 24551


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

gkalpak reviewed (approved):
I left a couple of comments. Otherwise lgtm.

##
Not directly related to this PR:

It seems we have similar (but not 100% the same) code in `@angular/animations`' [shared.ts][1] and `@angular/upgrade`'s [downgrade_component_adapter.ts][2]. I wonder if it would make sense to have a shared (internal) utility package for such helpers.

Also, these helpers are not 100% the same. It is hard to tell whether the differences account for slightly different needs of each package and how the helpers are used or whether we should align all implementations 😕 

[1]: https://github.com/angular/angular/blob/eba9dd364e51a6e857a94fd5f8980fec4350c856/packages/animations/browser/src/render/shared.ts#L168-L181
[2]: https://github.com/angular/angular/blob/eba9dd364e51a6e857a94fd5f8980fec4350c856/packages/upgrade/src/common/src/downgrade_component_adapter.ts#L291-L300

CaerusKaru commented on packages/elements/src/utils.ts:
```diff
@@ -8,9 +8,18 @@
-  return elProto.matches || elProto.matchesSelector || elProto.mozMatchesSelector ||
-      elProto.msMatchesSelector || elProto.oMatchesSelector || elProto.webkitMatchesSelector;
+  if (typeof Element === 'undefined' || Element.prototype.matches) {
+    return (element: any, selector: string) => element.matches(selector);
```
Since this is borrowed from animations shared code, I'm not entirely sure of the typing decisions. But I think until we consolidate, we should make sure the code is identical to avoid people trying to figure out which one is "better" later.

petebacondarwin commented on packages/elements/src/utils.ts:
```diff
@@ -7,11 +7,36 @@
+  // that Webpack polyfilled it by looking at `process.browser`, it's very Webpack-specific and
+  // might not be future-proof. Instead we look at the stringified version of `process` which
+  // is `[object process]` in Node and `[object Object]` when polyfilled.
+  return typeof process !== 'undefined' && {}.toString.call(process) === '[object process]';
```
I guess you are using `{}.toString.call()` because `process` could feasibly be `null`?

petebacondarwin commented on packages/elements/src/utils.ts:
```diff
@@ -7,11 +7,36 @@
+let _matches: (element: any, selector: string) => boolean = (element: any, selector: string) =>
+    false;
+
+if (_isNode || typeof Element !== 'undefined') {
+  _matches = (() => {
+    if (_isNode || Element.prototype.matches) {
+      return (element: any, selector: string) => element.matches(selector);
+    } else {
+      const proto = Element.prototype as any;
+      const fn = proto.matchesSelector || proto.mozMatchesSelector || proto.msMatchesSelector ||
+          proto.oMatchesSelector || proto.webkitMatchesSelector;
+      if (fn) {
+        return (element: any, selector: string) => fn.apply(element, [selector]);
+      } else {
+        return _matches;
+      }
+    }
+  })();
+}
```
This logic seems super complex. Is it to try to keep the code-size down or something else?

petebacondarwin replied:
I guess it is just a copy from the animations library.

alan-agius4 replied:
I also find this to be super complex.

petebacondarwin reviewed (approved):
Reviewed-for: size-tracking

Not going to block the merge. I guess 600 bytes is not a killer.

But I think it would be good to consider either (or both) of the following:

1. Make the `_matches` code easier to understand
2. Make the `_matches` code smaller

It is likely that these two are in opposition to each other. But I feel that as it stands we have the worst of both scenarios. I feel that it could be smaller or simpler...

Another option, as @gkalpak suggested, is to move this to a shared function, since I imagine that `elements` is rarely used without `animations`, in which case we are getting hit twice for this logic. Both of these have `core` as peer dependency, so it could just be moved there an exported "internally".

jelbourn reviewed (commented):
Would it be possible to add this shim exclusively to server environments rather than always baking it into the client-side lib?

alan-agius4 commented on packages/elements/src/utils.ts:
```diff
@@ -7,11 +7,36 @@
-  return elProto.matches || elProto.matchesSelector || elProto.mozMatchesSelector ||
-      elProto.msMatchesSelector || elProto.oMatchesSelector || elProto.webkitMatchesSelector;
-})();
+export function isNode(): boolean {
```
I don’t see any reason why this should be exported?

alan-agius4 reviewed (dismissed):
Can we change this implementation to something like: https://github.com/angular/angular/blob/d1ea1f4c7f3358b730b0d94e65b00bc28cae279c/packages/upgrade/src/common/src/downgrade_component_adapter.ts#L291-L300

it's much simpler, and should solve the issue since it's not using an IIFE.

PatrickJS commented:
if we use `Object.getPrototypeOf` we can keep the code the same e.g.
https://github.com/angular/angular/pull/38388/files#diff-97b1abb49de0e54f94eba3f03982ee1cR103

gkalpak commented:
@PatrickJS: I don't think we can (see https://github.com/angular/angular/pull/38388/files#r473900039).

@CaerusKaru: Are you still interested/available to move this forward? (If not, I'd be happy to take it over.)

CaerusKaru committed:
fix(elements): detect matchesSelector prototype without IIFE

Although in SSR we patch the global prototypes with DOM globals
like Element and Node, this patch does not occur before the
matches function is called in Angular Elements. This is similar
to the behavior in @angular/upgrade.

Fixes #24551

gkalpak commented on packages/elements/src/utils.ts:
```diff
@@ -96,11 +90,20 @@ export function kebabToCamelCase(input: string): string {
+
 /**
  * Check whether an `Element` matches a CSS selector.
+ * NOTE: this is duplicated from @angular/upgrade, and can
+ * be consolidated in the future
```
:100:

gkalpak reviewed (approved):
Still lgtm :+1:
Thx, @CaerusKaru!

atscott commented:
[presubmit](http://test/OCL:336295809:BASE:336295836:1602256688031:8b350869)

vinothbabu commented:
Does this issue fix the `createCustomElement` issue in SSR?

gkalpak commented:
What is the `createCustomElement()` issue? 😕

vinothbabu commented:
> What is the `createCustomElement()` issue? 😕

Angular Element is not available in SSR. (Below is the issue ticket). 
https://github.com/angular/angular/issues/24551

gkalpak commented:
Yes, this PR should have fixed #24551.

vinothbabu commented:
> Yes, this PR should have fixed #24551.

Wow, that's a great news. Would that mean we can use custom-elements in SSR and which release we are making it?

gkalpak commented:
It will be in the next 10.1.x release (10.1.6 - assuming there are going to be more v10 releases before v11) and the next 11.x release (11.0.0-next.6).
